### DISCUSSION TITLE
Generics and passing class name to base class

### DISCUSSION THREAD

**Discussion by gwww** [microsoft/pyright#3443] [Q&A] [answered]
I have something similar to the following code, which typing works and the code works (the example code below works). From a typing perspective, my questions are:

- Is this the "right" way?
- Are there other ways of doing this?
- Is it possible to remove passing the class name as a parameter (Python 3.9+)

```
from typing import Generic, Type, TypeVar

class Element:
    def __init__(self, index: int):
        print(f"Element init {index}")

T = TypeVar("T", bound=Element)

class Elements(Generic[T]):
    def __init__(self, class_: Type[T], how_many: int):
        print("Elements init")
        self.elements = [class_(i) for i in range(how_many)]

class Zone(Element):
    def __init__(self, index: int):
        print(f"Zone init {index}")
        super().__init__(index)

class Zones(Elements[Zone]):
    def __init__(self):
        print("Zones init")
        super().__init__(Zone, 5)

# A bunch more classes with similar pattern to Zone/Zones

zones = Zones()
```

Thanks!

**Comment by erictraut** [comment, marked as answer]
This pattern is a bit unusual in that it's combining a "container" pattern and a "factory" pattern. `Elements` is responsible for both constructing and containing a set of `Element` objects. More commonly, the job of constructing is distinct from the job of containing.

If your goal is to build a class that is responsible for constructing and containing, then this looks like a reasonable implementation.

Here's an alternative approach that moves the factory functionality out of the container:

```python
from typing import Generic, Sequence, TypeVar

class Element:
    def __init__(self, index: int): ...

    @classmethod
    def create_multiple(cls, how_many: int):
        return [cls(x) for x in range(how_many)]

class Zone(Element): ...

TElement = TypeVar("TElement", bound=Element)

class ElementContainer(Generic[TElement]):
    def __init__(self, elements: Sequence[TElement] | None = None):
        self.elements = [] if elements is None else list(elements)

class ZoneContainer(ElementContainer[Zone]): ...

# Create an empty container
zones1 = ZoneContainer()

# Manually construct list of elements to store in container
zones2 = ZoneContainer([Zone(i) for i in range(3)])

# Use class factory method
zones3 = ZoneContainer(Zone.create_multiple(4))
```

**Reply by gwww** [reply]
I appreciate your thoughtful answer! Thank you!

This has me thinking! I'm not sure to what end yet, but I'm pondering how to separate concerns which this does.

From a typing perspective and using the example code in my question, it feels like I have the typing right.
